File: Lib/odim_source.py
#!/usr/bin/env python
'''
Copyright (C) 2011- Swedish Meteorological and Hydrological Institute (SMHI)

This file is part of RAVE.

RAVE is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

RAVE is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with RAVE.  If not, see <http://www.gnu.org/licenses/>.
'''
## Various look-ups of the /what/source attribute

## @file
## @author Daniel Michelson, SMHI
## @date 2011-09-06

import sys
from rave_defines import UTF8, ODIM_SOURCE_FILE
import xml.etree.ElementTree as ET


## Dictionaries containing look-ups for identifiers except CTY.
# WMO, RAD, PLC, and SOURCE all use the NOD as the look-up
# NOD uses the WMO number as the look-up 
# CCCC will be the same for all radars from a given country, so there'll be
# a lot of redundancy, but this is needed to create Odyssey file strings.
NOD, WNOD, RNOD, WMO, RAD, PLC, CCCC, WIGOS = {None:None}, {None:None}, {None:None}, {None:None}, {None:None}, {None:None}, {None:None}, {None:None}
SOURCE = {None:None}

initialized = 0

# Hack to come around string handling differences between py3 and py27.
# We might consider go all-in on unicode handling but that will require some time.
use_source_encoding=False
if sys.version_info < (3,):
    use_source_encoding=True
    

## Initializer. Reads XML and puts the values into dictionaries.
# Because all radars must have a NOD identifier, the other identifiers are looked up
# based on NOD.
def init():
    global initialized
    if initialized: return
    
    O = ET.parse(ODIM_SOURCE_FILE)
    DB = O.getroot()
    for country in list(DB):
        cccc, org = country.attrib["CCCC"], country.attrib["org"]
        for radar in list(country):
            nod = radar.tag
            CCCC[nod] = cccc
            keys = radar.attrib.keys()
            if "wmo" in keys: 
                wmo = radar.attrib["wmo"]
                WMO[nod] = wmo
            else:
                wmo = None
            if "rad" in keys: 
                rad = radar.attrib["rad"]
                RAD[nod] = rad
            else:
                rad = None
            if "plc" in keys: 
                plc = radar.attrib["plc"]
                PLC[nod] = plc
            else:
                plc = None

            if "wigos" in keys:
                wigos = radar.attrib["wigos"]
                WIGOS[nod] = wigos
            else:
                wigos = None

            if wmo not in ("00000", None): 
                NOD[wmo] = nod
                SOURCE[nod] = u"NOD:%s" % nod
                if wmo: SOURCE[nod] += ",WMO:%s" % wmo
                if rad: SOURCE[nod] += ",RAD:%s" % rad
                if plc: SOURCE[nod] += ",PLC:%s" % plc
                if wigos: SOURCE[nod] += ",WIGOS:%s" % wigos
            else:
                SOURCE[nod] = u"NOD:%s" % nod
                if rad: SOURCE[nod] += ",RAD:%s" % rad
                if org and wmo!=None: SOURCE[nod] += ",ORG:%s" % org
                if plc: SOURCE[nod] += ",PLC:%s" % plc
                if wigos: SOURCE[nod] += ",WIGOS:%s" % wigos

            if wigos is not None:
                WNOD[wigos] = nod

            if rad is not None:
                RNOD[rad] = nod

    initialized = 1


init()


## One-time function, so read radar-db from text file.
# @param filename string of the input file to read
# @ param xmlfile string of the output XML file to write. Defaults to \ref rave_defines.ODIM_SOURCE_FILE 
def text2Element(filename, xmlfile=ODIM_SOURCE_FILE):
    import __builtin__
    from rave_IO import prettyprint

    E = R = None

    fd = open(filename)
    LINES = fd.readlines()
    LINES = LINES[0].split("\r")
    fd.close()

    ALL = ET.Element("radar-db", {"author" : "Daniel Michelson"})
    
    for L in LINES:
        l = L.split('\t')
        if len(l) == 3:
            if E:
                ALL.append(E)
            org, rad_prefix, country = l
            E = ET.Element(country, {"org":org})
        elif len(l) == 5:
            CCCC, nr, wmo, nod, plc = l
            E.attrib["CCCC"] = CCCC
            R = ET.Element(nod, {"wmo":wmo, "rad":"%s%s" % (rad_prefix,nr), 
                                 "plc":plc.decode('utf-8')})
            E.append(R)
        else:
            print("FAILED to process %s" % l)

    ALL.append(E)

    fd = __builtin__.open(xmlfile, 'w')
    sys.stdout = fd
    print("<?xml version='1.0' encoding='%s'?>" % UTF8)
    prettyprint(ALL)
    fd.close()
    sys.stdout = sys.__stdout__
    

class ODIM_Source:
    ## Initializer
    # @param src string containing a '/what/source' attribute
    def __init__(self, src=None):
        self.source = src
        if not isinstance(self.source,bytes) and self.source is not None:
          self.source = bytes(self.source, UTF8)
        self.wmo = self.nod = self.rad = self.plc = self.org = self.cty = self.cmt = self.wigos = None
        if self.source: self.split_source()

    ## Splits the input string into identifier values        
    def split_source(self):
        split = self.source.split(b',')
        for s in split:
            prefix, value = s.split(b':')
            prefix = prefix.lower()  # safety precaution in case someone changes case in their files
            if   prefix == b'wmo': self.wmo = value.decode(UTF8)  # Keep this as a string!
            elif prefix == b'rad': self.rad = value.decode(UTF8)
            elif prefix == b'plc': self.plc = value.decode(UTF8)
            elif prefix == b'nod': self.nod = value.decode(UTF8)
            elif prefix == b'org': self.org = value.decode(UTF8)
            elif prefix == b'cty': self.cty = value.decode(UTF8)
            elif prefix == b'cmt': self.cmt = value.decode(UTF8) 
            elif prefix == b'wigos': self.wigos = value.decode(UTF8) 

    def __str__(self):
        return "ODIM_Source(nod=%s, wmo=%s, rad=%s, plc=%s, org=%s, cty=%s, cmt=%s, wigos=%s" % \
            (self.nod or "", self.wmo or "", self.rad or "", self.plc or "", self.org or "", self.cty or "", self.cmt or "", self.wigos or "")

## Convenience function. Gets the NOD identifier from /what/source .
# Assumes that the NOD is there or can be looked up based on the WMO or WIGOS identifier.
# If neither can be found, then a 'n/a' (not available) is returned.
# @param obj input SCAN or PVOL object
# @return the NOD identifier or 'n/a'
def NODfromSource(obj):
    S = ODIM_Source(obj.source)
    if S.nod: 
        return S.nod
    elif S.wmo is not None and S.wmo in NOD:
        return NOD[S.wmo]
    elif S.wigos is not None and S.wigos in WNOD:
        return WNOD[S.wigos]

    return 'n/a'


## Convenience function. Checks and, if necessary, reformats a complete /what/source attribute .
# Doesn't return anything. 
# @param inobj input SCAN or PVOL object
def CheckSource(inobj):
    S = ODIM_Source(inobj.source)
    if not S.nod:
        if S.wmo is not None and S.wmo in NOD:
            S.nod = NOD[S.wmo]
        elif S.wigos is not None and S.wigos in WNOD:
            S.nod = WNOD[S.wigos]
        elif S.rad is not None and S.rad in RNOD:
            S.nod = RNOD[S.rad]
        else:
            return
        
        if SOURCE[S.nod] is not None:
            if use_source_encoding:
                inobj.source = SOURCE[S.nod].encode(UTF8)
            else:
                inobj.source = SOURCE[S.nod]


if __name__ == "__main__":
    pass

File: librave/toolbox/odim_source.c
/* --------------------------------------------------------------------
Copyright (C) 2025 Swedish Meteorological and Hydrological Institute, SMHI,

This file is part of RAVE.

RAVE is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

RAVE is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with RAVE.  If not, see <http://www.gnu.org/licenses/>.
------------------------------------------------------------------------*/
/**
 * Defines an odim source
 * @file
 * @author Anders Henja (Swedish Meteorological and Hydrological Institute, SMHI)
 * @date 2025-01-15
 */
#include "odim_source.h"
#include "rave_debug.h"
#include "rave_alloc.h"
#include <string.h>
#include <stdio.h>

/**
 * Represents the area
 */
struct _OdimSource_t {
  RAVE_OBJECT_HEAD /** Always on top */
  char* nod;     /**< the NOD */
  char* wmo;     /**< the WMO */
  char* wigos;   /**< the WIGOS */
  char* plc;     /**< the PLC */
  char* rad;     /**< the RAD */
  char* cccc;    /**< the CCCC */
  char* org;     /**< the ORG */
  char* source;  /**< the source string */
};

/*@{ Private functions */
/**
 * Constructor.
 */
static int OdimSource_constructor(RaveCoreObject* obj)
{
  OdimSource_t* this = (OdimSource_t*)obj;
  this->nod = NULL;
  this->wmo = NULL;
  this->wigos = NULL;
  this->plc = NULL;
  this->rad = NULL;
  this->cccc = NULL;
  this->org = NULL;
  this->source = NULL;
  return 1;
}

/**
 * Copy constructor
 */
static int OdimSource_copyconstructor(RaveCoreObject* obj, RaveCoreObject* srcobj)
{
  OdimSource_t* this = (OdimSource_t*)obj;
  OdimSource_t* src = (OdimSource_t*)srcobj;

  OdimSource_constructor(obj); // First just initialize everything like the constructor

  if (!OdimSource_setNod(this, src->nod)) {
    goto error;
  }
  if (!OdimSource_setWmo(this, src->wmo)) {
    goto error;
  }
  if (!OdimSource_setWigos(this, src->wigos)) {
    goto error;
  }
  if (!OdimSource_setPlc(this, src->plc)) {
    goto error;
  }
  if (!OdimSource_setRad(this, src->rad)) {
    goto error;
  }
  if (!OdimSource_setCccc(this, src->cccc)) {
    goto error;
  }
  if (!OdimSource_setOrg(this, src->org)) {
    goto error;
  }
  if (src->source != NULL) {
    this->source = RAVE_STRDUP(src->source);
    if (this->source == NULL) {
      goto error;
    }
  }
  return 1;
error:
  RAVE_FREE(this->nod);
  RAVE_FREE(this->wmo);
  RAVE_FREE(this->wigos);
  RAVE_FREE(this->plc);
  RAVE_FREE(this->rad);
  RAVE_FREE(this->cccc);
  RAVE_FREE(this->org);
  RAVE_FREE(this->source);
  return 0;
}

/**
 * Destroys the odim source
 * @param[in] obj - the the OdimSource_t instance
 */
static void OdimSource_destructor(RaveCoreObject* obj)
{
  OdimSource_t* this = (OdimSource_t*)obj;
  RAVE_FREE(this->nod);
  RAVE_FREE(this->wmo);
  RAVE_FREE(this->wigos);
  RAVE_FREE(this->plc);
  RAVE_FREE(this->rad);
  RAVE_FREE(this->cccc);
  RAVE_FREE(this->org);
  RAVE_FREE(this->source);
}
/*@} End of Private functions */

/*@{ Interface functions */
char* OdimSource_getIdFromOdimSource(const char* source, const char* id)
{
  char* result = NULL;
  if (source != NULL && id != NULL) {
    char* p = strstr(source, id);
    if (p != NULL) {
      int len = 0;
      char* pbrk = NULL;
      p += strlen(id);
      len = strlen(p);
      pbrk = strpbrk((const char*)p, ",");

      if (pbrk != NULL) {
        len = pbrk - p;
      }

      result = RAVE_MALLOC(sizeof(char) * (len + 1));
      if (result != NULL) {
        strncpy(result, p, len);
        result[len] = '\0';
      }
    }
  }
  return result;
}

char* OdimSource_getIdFromOdimSourceInclusive(const char* source, const char* id)
{
  char* result = NULL;
  if (source != NULL && id != NULL) {
    char* p = strstr(source, id);
    if (p != NULL) {
      int len = 0;
      char* pbrk = NULL;
      len = strlen(p);
      pbrk = strpbrk((const char*)p, ",");

      if (pbrk != NULL) {
        len = pbrk - p;
      }

      result = RAVE_MALLOC(sizeof(char) * (len + 1));
      if (result != NULL) {
        strncpy(result, p, len);
        result[len] = '\0';
      }
    }
  }
  return result;
}


OdimSource_t* OdimSource_create(const char* nod, const char* wmo, const char* wigos, const char* plc, const char* rad, const char* cccc, const char* org)
{
  OdimSource_t *result = NULL, *source = NULL;
  if (nod == NULL) {
    RAVE_ERROR0("Must specify at least nod in order to create a odim source");
    return NULL;
  }
  source = RAVE_OBJECT_NEW(&OdimSource_TYPE);
  if (source != NULL) {
    if (!OdimSource_setNod(source, nod)) {
      goto fail;
    }
    if (!OdimSource_setWmo(source, wmo)) {
      goto fail;
    }
    if (!OdimSource_setWigos(source, wigos)) {
      goto fail;
    }
    if (!OdimSource_setPlc(source, plc)) {
      goto fail;
    }
    if (!OdimSource_setRad(source, rad)) {
      goto fail;
    }
    if (!OdimSource_setCccc(source, cccc)) {
      goto fail;
    }
    if (!OdimSource_setOrg(source, org)) {
      goto fail;
    }
  }
  result = RAVE_OBJECT_COPY(source);
fail:
  RAVE_OBJECT_RELEASE(source);
  return result;
}

int OdimSource_setNod(OdimSource_t* self, const char* sid)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  RAVE_FREE(self->nod);
  RAVE_FREE(self->source);
  if (sid != NULL) {
    self->nod = RAVE_STRDUP(sid);
    if (self->nod == NULL) {
      RAVE_CRITICAL0("Failure when copying nod");
      return 0;
    }
  }
  return 1;
}

const char* OdimSource_getNod(OdimSource_t* self)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  return (const char*)self->nod;
}

int OdimSource_setWmo(OdimSource_t* self, const char* sid)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  RAVE_FREE(self->wmo);
  RAVE_FREE(self->source);
  if (sid != NULL) {
    self->wmo = RAVE_STRDUP(sid);
    if (self->wmo == NULL) {
      RAVE_CRITICAL0("Failure when copying wmo");
      return 0;
    }
  }
  return 1;
}

const char* OdimSource_getWmo(OdimSource_t* self)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  return (const char*)self->wmo;
}

int OdimSource_setWigos(OdimSource_t* self, const char* sid)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  RAVE_FREE(self->wigos);
  RAVE_FREE(self->source);
  if (sid != NULL) {
    self->wigos = RAVE_STRDUP(sid);
    if (self->wigos == NULL) {
      RAVE_CRITICAL0("Failure when copying wigos");
      return 0;
    }
  }
  return 1;
}

const char* OdimSource_getWigos(OdimSource_t* self)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  return (const char*)self->wigos;
}

int OdimSource_setPlc(OdimSource_t* self, const char* sid)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  RAVE_FREE(self->plc);
  RAVE_FREE(self->source);
  if (sid != NULL) {
    self->plc = RAVE_STRDUP(sid);
    if (self->plc == NULL) {
      RAVE_CRITICAL0("Failure when copying plc");
      return 0;
    }
  }
  return 1;
}

const char* OdimSource_getPlc(OdimSource_t* self)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  return (const char*)self->plc;
}

int OdimSource_setRad(OdimSource_t* self, const char* sid)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  RAVE_FREE(self->rad);
  RAVE_FREE(self->source);
  if (sid != NULL) {
    self->rad = RAVE_STRDUP(sid);
    if (self->rad == NULL) {
      RAVE_CRITICAL0("Failure when copying rad");
      return 0;
    }
  }
  return 1;
}

const char* OdimSource_getRad(OdimSource_t* self)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  return (const char*)self->rad;
}

int OdimSource_setCccc(OdimSource_t* self, const char* sid)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  RAVE_FREE(self->cccc);
  if (sid != NULL) {
    self->cccc = RAVE_STRDUP(sid);
    if (self->cccc == NULL) {
      RAVE_CRITICAL0("Failure when copying cccc");
      return 0;
    }
  }
  return 1;
}

const char* OdimSource_getCccc(OdimSource_t* self)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  return (const char*)self->cccc;
}

int OdimSource_setOrg(OdimSource_t* self, const char* sid)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  RAVE_FREE(self->org);
  if (sid != NULL) {
    self->org = RAVE_STRDUP(sid);
    if (self->org == NULL) {
      RAVE_CRITICAL0("Failure when copying org");
      return 0;
    }
  }
  return 1;
}

const char* OdimSource_getOrg(OdimSource_t* self)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  return (const char*)self->org;
}

const char* OdimSource_getSource(OdimSource_t* self)
{
  RAVE_ASSERT((self != NULL), "self == NULL");
  if (self->source == NULL) {
    char buff[1024];
    char* ptr = buff;
    int len = 1024, sz = 0;
    sz = snprintf(ptr, len, "NOD:%s", self->nod);
    ptr += sz;
    len -= sz;
    if (self->wmo != NULL && strcmp("00000", self->wmo) != 0) {
      sz = snprintf(ptr, len, ",WMO:%s", self->wmo);
      ptr += sz;
      len -= sz;
    }
    if (self->rad != NULL) {
      sz = snprintf(ptr, len, ",RAD:%s", self->rad);
      ptr += sz;
      len -= sz;
    }
    if (self->plc != NULL) {
      sz = snprintf(ptr, len, ",PLC:%s", self->plc);
      ptr += sz;
      len -= sz;
    }
    if (self->wigos != NULL) {
      sz = snprintf(ptr, len, ",WIGOS:%s", self->wigos);
      ptr += sz;
      len -= sz;
    }

    self->source = RAVE_STRDUP(buff);
  }
  return self->source;
}

/*@} End of Interface functions */

RaveCoreObjectType OdimSource_TYPE = {
    "OdimSource",
    sizeof(OdimSource_t),
    OdimSource_constructor,
    OdimSource_destructor,
    OdimSource_copyconstructor
};
